CTS1.rst

```diff
@@ -0,0 +1,666 @@
+
+  AccountAddress ::= (address: Byte³²)
+
+.. _CTS-ContractAddress:
+
+``ContractAddress``
+^^^^^^^^^^^^^^^^^^^
+
+An address of a contract instance.
+It consists of an index and a subindex both unsigned 64 bit integers.
+
+It is serialized as: first 8 bytes for the index (``index``) followed by 8 bytes for the subindex (``subindex``) both little endian::
+
+  ContractAddress ::= (index: Byte⁸) (subindex: Byte⁸)
+
+.. _CTS-Address:
+
+``Address``
+^^^^^^^^^^^
+
+Is either an account address or a contract address.
+
+It is serialized as: First byte indicates whether it is an account address or a contract address.
+In case the first byte is 0 then ``AccountAddress`` (``address``) is followed.
+In case the first byte is 1 then ``ContractAddress`` (``address``) is followed::
+
+  Address ::= (0: Byte) (address: AccountAddress)
+            | (1: Byte) (address: ContractAddress)
+
+
+.. _CTS-Receiver:
+
+``Receiver``
+^^^^^^^^^^^^
+
+The receiving address of a transfer, which is either an account address or a contract address.
+In the case of a contract address, additional information such as the name of a receive function and some additional bytes to call on the receiving contract.
+
+It is serialized as: First byte indicates whether it is an account address or a contract address.
+In case the first byte is 0 then ``AccountAddress`` (``address``) is followed.
+In case the first byte is 1 then ``ContractAddress`` (``address``), bytes for :ref:`CTS-ReceiveHookName` (``hook``) is followed::
+
+    Receiver ::= (0: Byte) (address: AccountAddress)
+               | (1: Byte) (address: ContractAddress) (hook: ReceiveHookName)
+
+.. _CTS-AdditionalData:
+
+``AdditionalData``
+^^^^^^^^^^^^^^^^^^^
+
+Additional bytes to include in a transfer, can be used to add additional parameters for the transfer function call.
+
+It is serialized as: the first 2 bytes encodes the length (``n``) of the data, followed this many bytes for the data (``data``)::
+
+  AdditionalData ::= (n: Byte²) (data: Byteⁿ)
+
+.. note::
+
+  This type is passed in a parameter for smart contract function calls, be aware of the parameter size limit of 1024 bytes.
+
+.. _CTS-functions:
+
+Contract functions
+------------------
+
+A smart contract implementing CTS1 MUST export three functions :ref:`CTS-functions-transfer`, :ref:`CTS-functions-updateOperator` and :ref:`CTS-functions-balanceOf` according to the following description:
+
+.. _CTS-functions-transfer:
+
+``transfer``
+^^^^^^^^^^^^
+
+Executes a list of token transfers.
+A transfer is a token ID, an amount of this token ID to be transferred from one address to some other address.
+
+When transferring tokens to a contract address additional information for a receive function hook to trigger is required.
+
+Parameter
+~~~~~~~~~
+
+The parameter is a list of transfers.
+
+It is serialized as: 1 byte representing the number of transfers (``n``) followed by the bytes for this number of transfers (``transfers``).
```

Comment:
I can't remember whether we discussed this, but is it not perhaps too limiting to mandate 1 byte for length?

If we relax the limits on the parameter size in the future this might be unnecessary.

Reply:
I don't think we have.

But for your information, the smallest possible transfer is 46 bytes, meaning the parameter size would have to be more than 11.7kB to be able to hit the limit of having one byte.

Reply:
Also since every transfer will result in an event being logged, the 64 log per contract function will also put a limit on this.